std/range/package.d

```diff
@@ -9594,6 +9594,514 @@ auto refRange(R)(R* range)
     foo(r);
 }
 
+/**
+Bitwise adapter over integral type ranges. Consumes the range elements bit by bit.
+*/
+struct Bitwise(R)
+    if (isInputRange!R && isIntegral!(ElementType!R))
+{
+private:
+    alias ElemType = ElementType!R;
+    alias UnsignedElemType = Unsigned!ElemType;
+
+    R parent;
+
+    enum bitsNum = ElemType.sizeof * 8;
+    size_t maskPos = bitsNum;
+
+    static if (hasLength!R)
+    {
+        ulong len;
+    }
+
+    static if (isBidirectionalRange!R)
+    {
+        size_t backMaskPos = 1;
+    }
+
+public:
+    this()(auto ref R range)
+    {
+        parent = range;
+
+        static if (hasLength!R)
+        {
+            len = parent.length * bitsNum;
+        }
+    }
+
+    bool empty()
+    {
+        static if (hasLength!R)
+        {
+            return len == 0;
+        }
+        else static if (isBidirectionalRange!R)
+        {
+            bool isOverlapping = parent.empty;
+
+            if (!parent.empty)
+            {
+                /*
+                   If we have consumed the last element of the range both from
+                   the front and the back, then the masks positions will overlap
+                 */
+                R parentCopy = parent.save;
+
+                parentCopy.popFront;
+                isOverlapping = parentCopy.empty && (maskPos < backMaskPos);
+            }
+
+            return parent.empty || isOverlapping;
+        }
+        else
+        {
+            /*
+               If we consumed the last element of the range, but not all the
+               bits in the last element
+             */
+            return parent.empty;
+        }
+    }
+
+    bool front()
+    {
+        assert(!empty());
+
+        return (parent.front & mask(maskPos)) != 0;
+    }
+
+    void popFront()
+    {
+        --maskPos;
+
+        if (maskPos == 0)
+        {
+            parent.popFront;
+            maskPos = bitsNum;
+        }
+
```

Comment:
too much whitespace - in Phobos we don't use whitespace around each compound statement